test-harness/specs/parameters-ac9.oas3.txt

```diff
@@ -0,0 +1,36 @@
+====test====
+When I send a request to an operation
+And the operation has a required path param
+And the path param has regex pattern
+And the request path does not match the pattern
+Then I should get a 422 response back
+And the payload of the response should be in application/problem+json content-type
+====spec====
+openapi: 3.0.0
+paths:
+  '/path/{param}':
+    parameters:
+      - name: param
+        in: path
+        required: true
```

Comment:
If a path param is not provided in this path `/foo/{bar}/baz` that would be `/foo//baz`? That would just be a 404 I guess?

Reply:
Good point, we can either take this as a validation error (missing required parameter) or return 404 as a missing route error. Hard to say for me which is the best.

Reply:
Prism returns 422 for `/foo//baz`. I just added a test case covering that.

1) 422 is a correct error for such case:
a) the path is known (thus not 404) however, the `bar` param is missing.
b) imagine that `bar` is defined as a number & request path = `/foo/2118/baz` - this also results in 422
c) having that said, invalid format and missing param are similar cases and should both lead to 422

2. On the other hand we have a case where param is last part of path: `/foo/{bar}`.
The input `/foo` results in 404 (test case: https://github.com/stoplightio/prism/blob/master/test-harness/specs/parameters-ac1.oas3.txt).

Even though 2) I think 1) is still correct behavior.

Reply:
It's probably just a personal opinion but I would expect `/foo//bar` to be a 404. Path parameters are a bit of a invented concept which doesnt really exist in HTTP, so... considering it an "invalid parameter" is mostly just down to us naming everything "validation" when really it is "looking for various types of failures are responding accordingly". 

What happens if you go to `https://github.com/stoplightio//pull/702`? 

For other failures, I think we could probably get some clarity here by checking common JavaScript routing libraries. Any routing system which allows a `/foo/:integer/bar`? If you send a string in as that param, does that 404 or 422? 

I think we consider too many things "validation", and that is confusing us when it comes to figuring out what prism should do.

Reply:
@StefanDywersant How hard would it be to treat this as a 404?

Reply:
The logic for checking `deprecated` and `required` is one place for all parameter types (query, body, header, and path). It would require some gymnastics to treat path-param-required differently. If I add that in this PR it will be soon redone during validators refactor. I'd do that inside validator-refactor PR to avoid doubling the work.

Additionally, I still have some doubts about whether it is a good approach:

`https://github.com/stoplightio//pull/702` - returns 404 because there is no such repo called ''. That's because it would not pass validation during repo creation.

Let's consider another example. Github repo-name-validation disallows '_'.

`https://github.com/stoplightio/_gigi_dagostino/pull/702` - shall it return 404 or 422?

I'm sure that it will respond with 404, again, the repo can't exist because it would not pass repo-name-validation. I'd say that 404 is a data-driven error and Prism has no database.


Summary:
`Prism pretends that it has some database and guesses whether a resource exists`
vs
`Prism focuses on request validation against the API description`

or

`Developers are lazy and don't validate input params if they don't need to, so Prism is`
vs
`Developers are lazy and don't validate input params if they don't need to, so Prism isn't`

Reply:
Ok I _think_ I understand what you're saying.

The basic idea would be that since Prism has the knowledge of your API it can proactively say "hey I know what you're trying to do — and you're missing a parameter so here's a nice 422 saying that this is not ok" — while a regular web server won't effectively know what's going on and has no choice than answering 404.

Interesting; on one side, we can provide a helpful error message to the user; on the other, we want to try to behave as much as possible as a real web server.

> It would require some gymnastics to treat path-param-required differently

Yeah I understand; moreover this kind of error should be thrown by the router component and not the deserialiser/validator. It's kind of a big effort right now and I think it's be better to wait until the validation stuff has been refactored.

I vote to keep (at least temporarily) the 422 status code, and reason about this in the future

Reply:
Agree! The moment of the decision whether prism is more-validation or more-mocking is coming :)